template/template.go

```diff
@@ -31,28 +31,94 @@ type Template struct {
 	locale   language.Tag
 }
 
-type variable struct {
-	name        *string
-	formatValue *string // TODO(mvilks): "selectValue" to be implemented
-	expression  ast.Expression
+// ResolvedValue keeps the result of the Expression resolution with optionally
+// defined format() and selectKey() functions for Format and Select contexts.
+type ResolvedValue struct {
+	value     any
+	selectKey func(keys []string) string
+	format    func() string
+	err       error
 }
 
-func newVariable(e ast.Expression, name *string) variable { return variable{expression: e, name: name} }
+func defaultFormat(value any) string {
+	switch v := value.(type) {
+	default:
+		s, err := castAs[string](v) // if underlying type is not string, return error
+		if err != nil {
+			return "" // errorf("unsupported value type: %T: %w", r.value, err)
+		}
+
+		return s
+	case fmt.Stringer:
+		return v.String()
+	case string, []byte, []rune, int, int8, int16, int32, int64,
+		uint, uint8, uint16, uint32, uint64, float32, float64, bool,
+		complex64, complex128, error:
+		return fmt.Sprint(v)
+	}
+}
+
+func defaultSelectKey(value any, keys []string) string {
+	v := defaultFormat(value)
+
+	for _, k := range keys {
+		if v == k {
+			return k
+		}
+	}
+
+	return ast.CatchAllKey{}.String()
+}
+
+// String makes the ResolvedValue implement the fmt.Stringer interface.
+func (r *ResolvedValue) String() string {
+	if r.format != nil {
+		return r.format()
+	}
+
+	return defaultFormat(r.value)
+}
+
+// OptionalFunction is a function to apply to the ResolvedValue.
+type OptionalFunction func(*ResolvedValue)
```

Comment:
It is an option, not optional. Also, I'd like to suggest the following name to reflect its purpose.

```suggestion
type ResolvedValueOpt func(*ResolvedValue)
```